packages/cli/src/cli.ts

```diff
@@ -53,14 +53,18 @@ export async function cli(args: string[] = process.argv) {
           .demandCommand();
       },
       handler: (_args) => {},
+    })
+    .fail(async (msg, err, yargs) => {
+      const Consola = await import("consola");
+      if (err && err instanceof ExitProcessError) {
+        Consola.consola.error(err);
+      } else {
+        yargs.showHelp();
+        // eslint-disable-next-line no-console
+        console.error(msg);
+      }
+      process.exit(1);
```

Comment:
- If err is defined but not ExitProcessError should we follow the yargs example and rethrow the error with stacktrace? These should be uncaught runtime exceptions given the ExitProcessError special casing is msg defined in these cases?
- What are the cases in which err is not defined is that for built in validation errors from yargs? Is it desirable to show help before or after the message? The yargs examples seem to default to printing help after the message

Reply:
> What are the cases in which err is not defined is that for built in validation errors from yargs? Is it desirable to show help before or after the message? The yargs examples seem to default to printing help after the message

Yup it seems to be for built in validation errors. The default out of the box case is to actually showHelp then print 
the message.  I prefer the error message to be shown last as that is more or less the expectation when faced with an error.

> If err is defined but not ExitProcessError should we follow the yargs example and rethrow the error with stacktrace? These should be uncaught runtime exceptions given the ExitProcessError special casing is msg defined in these cases?

I am not sure about this one, because the Errors thrown during the `check` will be in that category and for those I think we don't want to show the stacktrace because it doesn't add value to the user. So if we did want to show stacktrace to runtime exceptions we can throw a different kind of Error inside of `check` perhaps define a `CheckError`.

This is what this would look like for passing an invalid foundryUrl:

<img width="1057" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/62106d36-093a-4e09-953e-9000c2ea219f">

Reply:
What happens if you return string rather than an error in `.check()` out of interest? Does yargs throw that in a wrapped error or does it populate msg?
https://yargs.js.org/docs/#api-reference-checkfn-globaltrue

Reply:
Sweet, I played around and yargs returns the string in place of `err` object. So adding to the else part of fail this:

- String returned from check (Show help + returned string) OR validation error:
<img width="635" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/318c1a8a-9df3-4fae-a0c8-cd10610488f0">
<img width="610" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/fd32f41d-7a51-4731-8b44-2552578d5a70">

- ExitProcessorError (no help + Error and stacktrace with consola )
<img width="599" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/a17233ae-2964-4e7a-bde3-2721d461b015">

- runtime exception
<img width="899" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/32c14a58-19e8-4b5f-9bd9-b719e8ec829f">

Reply:
There is a 4th kind where any Error kind is thrown before yargs does it's parsing for example I changed the ConfigFile loading to throw ExitProcessError but that is handled outside of the `.fail` because yargs hasn't reached that part yet.

<img width="707" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/205de6b0-f88e-43c9-9f3b-9ca6e7fc1193">


But then if I re-add the original try/catch around parseAsync then I can special handle ExitProcessError again 👍 
<img width="656" alt="image" src="https://github.com/palantir/osdk-ts/assets/35533361/8e6e28d3-1617-4b53-a52c-a8a48faf569b">

Reply:
That way I think we covered all possible handled/unhandled exceptions/checks within and outside yargs